File: javascript/packages/config/test/misnamed-config.test.ts
import { describe, test, expect, beforeEach, afterEach, vi } from "vitest"
import { mkdirSync, writeFileSync, rmSync, existsSync } from "fs"
import { join } from "path"
import { tmpdir } from "os"

import { Config } from "../src/config.js"

describe("misnamed config files", () => {
  let testDir: string

  beforeEach(() => {
    testDir = join(tmpdir(), `herb-misnamed-config-test-${Date.now()}-${Math.random().toString(36).slice(2)}`)
    mkdirSync(join(testDir, ".git"), { recursive: true })
    writeFileSync(join(testDir, ".git", "HEAD"), "ref: refs/heads/main\n")
  })

  afterEach(() => {
    vi.restoreAllMocks()

    if (existsSync(testDir)) {
      rmSync(testDir, { recursive: true, force: true })
    }
  })

  function captureStderr() {
    const messages: string[] = []

    vi.spyOn(console, "error").mockImplementation((...args: any[]) => {
      messages.push(args.join(" "))
    })

    return messages
  }

  describe("Config.misnamedConfigPaths", () => {
    test("covers the wrong extension and the missing leading dot", () => {
      expect(Config.misnamedConfigPaths).toEqual([".herb.yaml", "herb.yml", "herb.yaml"])
    })
  })

  describe("Config.isMisnamedConfigPath", () => {
    test("recognizes misnamed config files", () => {
      expect(Config.isMisnamedConfigPath(".herb.yaml")).toBe(true)
      expect(Config.isMisnamedConfigPath("herb.yml")).toBe(true)
      expect(Config.isMisnamedConfigPath("herb.yaml")).toBe(true)
      expect(Config.isMisnamedConfigPath("/project/.herb.yaml")).toBe(true)
    })

    test("ignores the config file Herb reads", () => {
      expect(Config.isMisnamedConfigPath(".herb.yml")).toBe(false)
      expect(Config.isMisnamedConfigPath("/project/.herb.yml")).toBe(false)
    })

    test("ignores unrelated files", () => {
      expect(Config.isMisnamedConfigPath("config.yml")).toBe(false)
      expect(Config.isMisnamedConfigPath("myherb.yml")).toBe(false)
      expect(Config.isMisnamedConfigPath(".herb.yml.bak")).toBe(false)
      expect(Config.isMisnamedConfigPath("/project")).toBe(false)
    })
  })

  describe("Config.findMisnamedConfigPaths", () => {
    test("returns every misnamed config file", async () => {
      writeFileSync(join(testDir, ".herb.yaml"), "")
      writeFileSync(join(testDir, "herb.yaml"), "")

      expect(await Config.findMisnamedConfigPaths(testDir)).toEqual([
        join(testDir, ".herb.yaml"),
        join(testDir, "herb.yaml")
      ])
    })

    test("ignores the config file Herb reads", async () => {
      writeFileSync(join(testDir, ".herb.yml"), "")

      expect(await Config.findMisnamedConfigPaths(testDir)).toEqual([])
    })

  })

  describe("Config.misnamedConfigWarning", () => {
    test("names the file and the fix", () => {
      expect(Config.misnamedConfigWarning("/project/.herb.yaml")).toBe(
        "⚠ Ignoring /project/.herb.yaml: Herb only reads `.herb.yml`. Rename it to `.herb.yml` to apply it."
      )
    })
  })

  describe("Config.load", () => {
    test("warns about a config file with the wrong extension", async () => {
      writeFileSync(join(testDir, ".herb.yaml"), "linter:\n  enabled: false\n")

      const messages = captureStderr()

      await Config.load(testDir, { version: "0.10.3" })

      expect(messages).toContain(Config.misnamedConfigWarning(join(testDir, ".herb.yaml")))
    })

    test("warns about a config file missing the leading dot", async () => {
      writeFileSync(join(testDir, "herb.yml"), "linter:\n  enabled: false\n")

      const messages = captureStderr()

      await Config.load(testDir, { version: "0.10.3" })

      expect(messages).toContain(Config.misnamedConfigWarning(join(testDir, "herb.yml")))
    })

    test("warns once about every misnamed config file", async () => {
      for (const filename of Config.misnamedConfigPaths) {
        writeFileSync(join(testDir, filename), "")
      }

      const messages = captureStderr()

      await Config.load(testDir, { version: "0.10.3" })

      const warnings = messages.filter(message => message.includes("Herb only reads"))

      expect(warnings).toEqual(Config.misnamedConfigPaths.map(filename => Config.misnamedConfigWarning(join(testDir, filename))))
    })

    test("warns when a misnamed config file is passed explicitly", async () => {
      const misnamedPath = join(testDir, ".herb.yaml")

      writeFileSync(misnamedPath, "linter:\n  enabled: false\n")

      const messages = captureStderr()

      await Config.load(misnamedPath, { version: "0.10.3" })

      expect(messages.filter(message => message.includes("Herb only reads"))).toEqual([Config.misnamedConfigWarning(misnamedPath)])
    })

    test("does not warn when only the config file Herb reads exists", async () => {
      writeFileSync(join(testDir, ".herb.yml"), "linter:\n  enabled: false\n")

      const messages = captureStderr()

      await Config.load(testDir, { version: "0.10.3" })

      expect(messages.filter(message => message.includes("Herb only reads"))).toEqual([])
    })

    test("does not warn in silent mode", async () => {
      writeFileSync(join(testDir, ".herb.yaml"), "")

      const messages = captureStderr()

      await Config.load(testDir, { version: "0.10.3", silent: true })

      expect(messages).toEqual([])
    })

    test("does not read a misnamed config file", async () => {
      writeFileSync(join(testDir, ".herb.yaml"), "version: 0.8.0\nformatter:\n  indentWidth: 8\n")

      const config = await Config.load(testDir, { version: "0.10.3", silent: true })

      expect(config.configVersion).toBeUndefined()
      expect(config.formatter?.indentWidth).not.toBe(8)
    })

    test("still reads the config file next to a misnamed one", async () => {
      writeFileSync(join(testDir, ".herb.yml"), "version: 0.8.0\nformatter:\n  indentWidth: 4\n")
      writeFileSync(join(testDir, ".herb.yaml"), "version: 0.9.0\nformatter:\n  indentWidth: 8\n")

      const messages = captureStderr()

      const config = await Config.load(testDir, { version: "0.10.3" })

      expect(config.path).toBe(join(testDir, ".herb.yml"))
      expect(config.configVersion).toBe("0.8.0")
      expect(config.formatter?.indentWidth).toBe(4)
      expect(messages).toContain(Config.misnamedConfigWarning(join(testDir, ".herb.yaml")))
    })

    test("creates the default config next to a misnamed one instead of exiting", async () => {
      writeFileSync(join(testDir, ".herb.yaml"), "linter:\n  enabled: false\n")

      const messages = captureStderr()

      await Config.load(testDir, { version: "0.10.3", createIfMissing: true })

      expect(existsSync(join(testDir, ".herb.yml"))).toBe(true)
      expect(messages).toContain(Config.misnamedConfigWarning(join(testDir, ".herb.yaml")))
    })
  })

  describe("Config.validateConfigText", () => {
    test("reports a config file with the wrong extension", async () => {
      writeFileSync(join(testDir, ".herb.yaml"), "")

      const errors = await Config.validateConfigText("version: 0.10.3\n", { version: "0.10.3", projectPath: testDir })
      const warning = errors.find(error => error.code === "wrong_file_extension")

      expect(warning?.message).toBe("Found .herb.yaml file. Please rename to .herb.yml")
      expect(warning?.severity).toBe("warning")
    })

    test("reports a config file missing the leading dot", async () => {
      writeFileSync(join(testDir, "herb.yml"), "")

      const errors = await Config.validateConfigText("version: 0.10.3\n", { version: "0.10.3", projectPath: testDir })
      const warning = errors.find(error => error.code === "wrong_file_extension")

      expect(warning?.message).toBe("Found herb.yml file. Please rename to .herb.yml")
    })

    test("reports every misnamed config file", async () => {
      for (const filename of Config.misnamedConfigPaths) {
        writeFileSync(join(testDir, filename), "")
      }

      const errors = await Config.validateConfigText("version: 0.10.3\n", { version: "0.10.3", projectPath: testDir })

      expect(errors.filter(error => error.code === "wrong_file_extension")).toHaveLength(3)
    })

    test("does not report anything when only the config file Herb reads exists", async () => {
      writeFileSync(join(testDir, ".herb.yml"), "")

      const errors = await Config.validateConfigText("version: 0.10.3\n", { version: "0.10.3", projectPath: testDir })

      expect(errors.filter(error => error.code === "wrong_file_extension")).toEqual([])
    })
  })
})

File: rust/herb-config/tests/misnamed_config_test.rs
use std::fs;
use std::path::Path;

use herb_config::{Config, MISNAMED_CONFIG_PATHS};

#[test]
fn misnamed_config_paths_covers_the_extension_and_the_missing_leading_dot() {
  assert_eq!(MISNAMED_CONFIG_PATHS, &[".herb.yaml", "herb.yml", "herb.yaml"]);
}

#[test]
fn is_misnamed_config_path_recognizes_misnamed_config_files() {
  assert!(Config::is_misnamed_config_path(Path::new(".herb.yaml")));
  assert!(Config::is_misnamed_config_path(Path::new("herb.yml")));
  assert!(Config::is_misnamed_config_path(Path::new("herb.yaml")));
  assert!(Config::is_misnamed_config_path(Path::new("/project/.herb.yaml")));
}

#[test]
fn is_misnamed_config_path_ignores_the_real_config_file() {
  assert!(!Config::is_misnamed_config_path(Path::new(".herb.yml")));
  assert!(!Config::is_misnamed_config_path(Path::new("/project/.herb.yml")));
}

#[test]
fn is_misnamed_config_path_ignores_unrelated_yaml_files() {
  assert!(!Config::is_misnamed_config_path(Path::new("config.yml")));
  assert!(!Config::is_misnamed_config_path(Path::new("myherb.yml")));
  assert!(!Config::is_misnamed_config_path(Path::new(".herb.yml.bak")));
}

#[test]
fn find_misnamed_config_paths_returns_every_misnamed_config_file() {
  let dir = tempfile::tempdir().unwrap();

  fs::write(dir.path().join(".herb.yaml"), "").unwrap();
  fs::write(dir.path().join("herb.yaml"), "").unwrap();

  let found = Config::find_misnamed_config_paths(dir.path());

  assert_eq!(found, vec![dir.path().join(".herb.yaml"), dir.path().join("herb.yaml")]);
}

#[test]
fn find_misnamed_config_paths_ignores_the_real_config_file() {
  let dir = tempfile::tempdir().unwrap();

  fs::write(dir.path().join(".herb.yml"), "").unwrap();

  assert!(Config::find_misnamed_config_paths(dir.path()).is_empty());
}

#[test]
fn misnamed_config_warning_names_the_file_and_the_fix() {
  let warning = Config::misnamed_config_warning(Path::new("/project/.herb.yaml"));

  assert_eq!(
    warning,
    "\u{26a0} Ignoring /project/.herb.yaml: Herb only reads `.herb.yml`. Rename it to `.herb.yml` to apply it."
  );
}

#[test]
fn load_still_uses_the_real_config_file_next_to_a_misnamed_one() {
  let dir = tempfile::tempdir().unwrap();

  fs::write(dir.path().join(".herb.yml"), "version: 0.10.3\nformatter:\n  indentWidth: 4\n").unwrap();
  fs::write(dir.path().join(".herb.yaml"), "version: 0.10.3\nformatter:\n  indentWidth: 8\n").unwrap();

  let config = Config::load(dir.path(), None).unwrap();

  assert_eq!(config.path.file_name().unwrap(), ".herb.yml");
  assert_eq!(config.formatter().unwrap().indent_width, Some(4));
}

#[test]
fn load_ignores_a_misnamed_config_file_and_falls_back_to_defaults() {
  let dir = tempfile::tempdir().unwrap();

  fs::write(dir.path().join("Gemfile"), "").unwrap();
  fs::write(dir.path().join(".herb.yaml"), "version: 0.10.3\nformatter:\n  indentWidth: 8\n").unwrap();

  let config = Config::load(dir.path(), None).unwrap();

  assert_eq!(config.config_version, None);
  assert_ne!(config.formatter().unwrap().indent_width, Some(8));
}

#[cfg(feature = "yerba")]
#[test]
fn load_for_cli_creates_the_default_config_next_to_a_misnamed_one() {
  let dir = tempfile::tempdir().unwrap();

  fs::write(dir.path().join("Gemfile"), "").unwrap();
  fs::write(dir.path().join(".herb.yaml"), "version: 0.10.3\n").unwrap();

  let config = Config::load_for_cli(dir.path(), None, true).unwrap();

  assert_eq!(config.path.file_name().unwrap(), ".herb.yml");
  assert!(dir.path().join(".herb.yml").exists());
}
